### ISSUE TITLE
More issues with option 'failglob'

### ISSUE THREAD

**Issue by Maelan** [scop/bash-completion#324]
I noticed several more issues with Bash’s option `failglob`, similar to the one described in issue #306. Again, the error message `bash: no match: …` is emitted, interlacing with the command line.

  * architecture x86-64
  * Linux 5.0.3-arch1-1-ARCH (Archlinux)
  * Bash 5.0.7(1)-release (from Archlinux’ package)
  * bash-completion 2.9.1 (from Archlinux’ package), and also with the most recent development version (commit 3702ae0)

### Bug 1: `_filedir_xspec` when `failglob` is set and the word starts with a tilde

This time, the affected completion function is `_filedir_xspec` (I initially noticed it with command `vim`, then tried it with other commands that use the same completion function), whose filename completion becomes buggy when the word being completed starts with a tilde: not only does it output the error message, but also it restricts to directory names.

```bash
$ env -i bash --norc --noprofile
bash-5.0$ source /usr/local/share/bash-completion/bash_completion
bash-5.0$ shopt -s failglob
bash-5.0$ ls -F ~/TEST
ADIR/  AFILE
bash-5.0$ vim TEST/A‹tab›‹tab›
ADIR/  AFILE  
bash-5.0$ vim ~/TEST/A‹tab›[bash: no match: \~/TEST/A
]DIR/^C
bash-5.0$ 
```

I’m joining a debug output (see next message).

### Bug 2: `ssh` when `failglob` is set and known_hosts contains an IP with angle brackets

Another buggy completion is that of command `ssh` (and related commands such as `scp` and `rsync`). When `failglob` is set and `~/.ssh/known_hosts` contains an IP address of the form `[123.45.67.89]:1234`, then no matter what the word being typed is, the completion issues the same error message with that IP address, and discards the command line being typed.

```bash
$ env -i bash --norc --noprofile
bash-5.0$ source /usr/local/share/bash-completion/bash_completion
bash-5.0$ ssh ‹tab›
(this suggests hostnames and IP address from ~/.ssh/config and ~/.ssh/known_hosts)
bash-5.0$ shopt -s failglob
bash-5.0$ ssh ‹tab›[bash: no match: [123.45.67.89]:1234]
bash-5.0$ 
```

Again I am joining a debug output (see next message).

**Comment by scop** [comment]
Yes, there probably are a lot of these issues because supporting failglob mode has never been a goal in any way. Ditto nounset FWIW. Not saying they shouldn't, just stating the current state of affairs. I'm not saying that they will be fixed either though, some of changes required for them may result in worse code readability so it's a tradeoff.

Aside, please consider using attachments instead of pasting large debug logs etc in the future, thanks.

**Comment by Maelan** [comment]
[debug1.txt](https://github.com/scop/bash-completion/files/3244795/debug1.txt)
[debug2.txt](https://github.com/scop/bash-completion/files/3244796/debug2.txt)
Ah, didn’t know about attachments, thanks.

**Comment by Maelan** [comment]
Fair enough, in this case maybe the README should mention `failglob` (and `nullglob`, as I also discovered more bugs with `_filedir_xspecs` and this option), in the same way as it mentions `nounset`.

I hacked an ugly workaround in my Bash config file:

```bash
# Workaround for bash-completion not supporting some Bash options.
# We disable the unsupported options when typing a command line, and re-enable
# them before Bash expands and executes the command line.
function save_shell_options() {
	_saved_set="$(set +o)"
	_saved_shopt="$(shopt -p nullglob failglob)"
	set +o nounset
	shopt -u nullglob failglob
}
function restore_shell_options() {
	[ -z "${_saved_set+x}" ] && return
	eval "$_saved_set"
	eval "$_saved_shopt"
	unset _saved_set _saved_shopt
}
# We use the PROMPT_COMMAND var to run a command just before showing the prompt.
# Note: For some reason, changes to shell options only take effect if this is
# the *last* command executed by PROMPT_COMMAND.
PROMPT_COMMAND="$PROMPT_COMMAND
save_shell_options;"
# We use the DEBUG trap to run a command after the command line is entered, but
# before it is expanded by Bash.
# Note: This is very hackish. The manual is not explicit as whether the trap is
# actually called *before* expansion, even though that’s what happens in
# practice. Moreover, the trap is called before each *simple* command, including
# before commands in PROMPT_COMMAND.
trap="$(trap -p DEBUG)" ; trap="${trap#trap -- }" ; trap="${trap% DEBUG}"
eval "trap 'restore_shell_options;'$trap DEBUG"
unset trap
# We cannot use PS1 and PS0, as we would only have access to a subshell.
```

**Comment by lucaswerkmeister** [comment]
I humbly suggest checking whether this issue still exists in the latest version – I wasn’t quite able to reproduce it myself before I upgraded, but going from bash-completion 2.11 to 2.14 seemed to resolve several other `failglob`-related issues I encountered (including in `ssh` completion), and it seems like there were several globbing-related refactorings in recent releases (such as 6981c03f3e, first included in 2.12). IMHO it’s quite possible that this issue was resolved as part of that.

**Comment by scop** [comment]
Thanks for reminding about this issue, @lucaswerkmeister. Bunch of work has indeed gone into making things work with failglob.

If @akinomyoga agrees, I think we could expand our CONTRIBUTING.md entry that says
> We want our completions to work in `posix` and `nounset` modes.

...to also mention failglob. To me that would be enough to close this issue -- we could leave failglob in the test suite off still if flipping it on sounds like too big a leap at this point. Refs #1213

**Comment by akinomyoga** [comment]
I agree on officially declaring support for `failglob` and requiring it in `CONTRIBUTING.md`.

I confirmed that the current test set passes with `failglob` without any modifications (actually, some of the tests needed to be fixed as in #1213).

I think we can close this issue now. There may still be some cases not covered by our tests, but we can process them when they turn out. We may also want to have a CI configuration with `failglob`, but there's an independent issue for that, #602, anyway.

**Comment by Maelan** [comment]
I confirm that I could still reproduce this (with a very slightly different failing behavior) with bash-completion 2.11-3, and that I cannot reproduce it anymore with bash-completion 2.14.0-2 (Bash 5.2.26), as provided by Archlinux packages.

Thank you both (@scop and @akinomyoga) for this! It’s scary that array splitting undergoes path expansion… Incidentally, with [all that work you’ve done around `_comp_compgen`](https://github.com/scop/bash-completion/projects/3#card-88207676) to avoid path expansions, does that mean that `~/` will stop from getting expanded to `/home/me/` when tab-completing? (that would be another relief :-) )

**Comment by scop** [comment]
>  if it implies that tab-completing will not expand ~/ or symbolic links anymore, that would be another relief. :-)

Re expanding `~/`, that's might be a different thing, and we still explicitly do that here and there `git grep -wF _comp_expand`, and see reasoning in the `_comp_expand` commentary. I'm not sure if all those would be necessary though.

Re expanding symlinks, I'm not sure what that refers to. Could you elaborate in a separate issue if you think we're doing something we should not?

**Comment by scop** [comment]
> I'm not sure if all those would be necessary though.

...and ones that are would deserve a comment explaining why, and a test suite case that demonstrates the failure unless there already is one. Custom filename stuff in scp and friends might be the reasoning for those, but hard to tell without a comment.

**Comment by Maelan** [comment]
(This is becoming off-topic.)

Apologies, that was a vague remembrance. I have now been able to reproduce the bug I had in mind (and that I was hitting daily) and, after a bit of investigation, I have been able to pinpoint it to Bash option `direxpand` (`shopt -s direxpand`). bash-completion itself does not seem to blame, because I also observe the buggy behavior with plain Bash, without bash-completion loaded. Perhaps bash-completion would want to secure itself against this bug of Bash, but it does not seem urgent.

For more detail: if tab-completing a path that contains a space, and there is at least one matching completion, and `direxpand` is set, then user input is rewritten in some unexpected ways, i.e. relative is made absolute and tilde is expanded (e.g. `ls dir\ with\ spaces/<tab>` → `ls /home/me/dir\ with\ spaces/` ; `ls ~/dir\ with\ spaces/<tab>` → `ls /home/me/dir\ with\ spaces/`). This also happens when the path contains a Bash variable (e.g. `ls ../$ME/Documents/<tab>` → `/home/me/Documents/`). The Bash manpage says:

> **direxpand**  
> If set, bash replaces directory names with the results of **word expansion** when performing filename completion. This changes the contents of the readline editing buffer. If not set, bash attempts to preserve what the user typed.

I am unable to find what “word expansion” means in this context, but I believe it does not include resolving relative paths to absolute paths, and it possibly does not include tilde expansion neither. In any case, the observed behavior is pretty much a bug because there is no reason to treat directories that contain spaces differently from directories that don’t.

Fortunately, the fix seems simple: just disable the option. The reason for me to enable it in the first place seems to be gone in recent versions of Bash (trying Bash 5.2.26 here) which apparently, is now able to complete on stuff like `$HOME/` or `../$ME/` without even expanding the variable in user input (and this works with bash-completion, too).
